### PR TITLE
Implement Entitymanager and populate security artifacts

### PR DESCRIPTION
**Pull request by gmacf**
This PR implements some basic entitymanager functionality and also reads the keypairs and OVs from disk.

### CODE REVIEW

File: server/entitymanager/entitymanager.go [hunk 1]
```diff
@@ -2,38 +2,154 @@
...
+		BootMode: bootMode,
+	}
+	m.chassisInventory[l] = &e
+	return m
 }
 
 // New returns a new in-memory entity manager.
-func New() *InMemoryEntityManager {
-	return &InMemoryEntityManager{}
+func New(artifacts *service.SecurityArtifacts) *InMemoryEntityManager {
```
Comment (marcushines) [inline server/entitymanager/entitymanager.go:149]: we should have a clear workflow on how artifacts get loaded into the entity manager - i am seeing AddChasssis and AddControlCard

File: server/entitymanager/entitymanager.go [hunk 2]
```diff
@@ -2,38 +2,162 @@
...
+		}
+		log.Infof("control card %v changed status from %v to %v", c.GetSerialNumber(), previousStatus, c.GetStatus())
+		m.controlCardStatuses[c.GetSerialNumber()] = c.GetStatus()
+	}
+	return nil
+}
+
+// Sign unmarshals the SignedResponse bytes then generates a signature from its Ownership Certificate private key.
+func (m *InMemoryEntityManager) Sign(resp *bootz.GetBootstrapDataResponse, serial string) error {
+	// Sign the response
```
Comment (sam-homa) [inline server/entitymanager/entitymanager.go:93]: nit: these comments are not providing much value imo.

File: server/entitymanager/entitymanager.go [hunk 3]
```diff
@@ -2,38 +2,162 @@
...
+	}
+	if resp.GetSignedResponse() == nil {
+		return status.Errorf(codes.InvalidArgument, "empty signed response")
+	}
+	signedResponseBytes, err := proto.Marshal(resp.GetSignedResponse())
+	if err != nil {
+		return err
+	}
+	hashed := sha256.Sum256(signedResponseBytes)
+	sig, err := rsa.SignPKCS1v15(nil, priv, crypto.SHA256, hashed[:])
```
Comment (sam-homa) [inline server/entitymanager/entitymanager.go:110]: I think ECs are becoming more common and so I think it'd be a good idea to somehow detect the priv key type and support both algorithms. could be done as an improvement in a later PR though.

File: server/entitymanager/entitymanager.go [hunk 4]
```diff
@@ -2,38 +2,162 @@
...
+	}
+	return nil
+}
+
+// Sign unmarshals the SignedResponse bytes then generates a signature from its Ownership Certificate private key.
+func (m *InMemoryEntityManager) Sign(resp *bootz.GetBootstrapDataResponse, serial string) error {
+	// Sign the response
+	block, _ := pem.Decode([]byte(m.artifacts.OC.Key))
+	if block == nil {
+		return status.Errorf(codes.Internal, "unable to decode OC private key")
```
Comment (sam-homa) [inline server/entitymanager/entitymanager.go:96]: The term "private key" is a little confusing to me. don't we only send public key in the OC (as a x509)?  Then the function argument is "priv", so I might be misunderstanding the flow here.

File: server/entitymanager/entitymanager_test.go [hunk 5]
```diff
@@ -0,0 +1,291 @@
...
+					t.Skip()
+				}
+				t.Errorf("Sign() err = %v, want %v", err, test.wantErr)
+			}
+			signedResponseBytes, err := proto.Marshal(test.resp.GetSignedResponse())
+			if err != nil {
+				t.Fatal(err)
+			}
+			hashed := sha256.Sum256(signedResponseBytes)
+			sigDecoded, err := base64.StdEncoding.DecodeString(test.resp.GetResponseSignature())
```
Comment (sam-homa) [inline server/entitymanager/entitymanager_test.go:159]: nit: it feels to me that sign and verifySign should be in a package that could be used by both server and client code and unit tests

File: server/entitymanager/entitymanager.go [hunk 6]
```diff
@@ -25,6 +26,7 @@ type InMemoryEntityManager struct {
 	// represents the current status of known control cards
 	controlCardStatuses map[string]bootz.ControlCardState_ControlCardStatus
 	artifacts           *service.SecurityArtifacts
+	mu                  sync.Mutex
```
Comment (marcushines) [inline server/entitymanager/entitymanager.go:30]: nit group the vars that mu protects below the mu

File: server/entitymanager/entitymanager_test.go [hunk 7]
```diff
@@ -0,0 +1,291 @@
...
+	}
+}
+
+func TestResolveChassis(t *testing.T) {
+	tests := []struct {
+		desc    string
+		input   *service.EntityLookup
+		want    *service.ChassisEntity
+		wantErr bool
+	}{
```
Comment (marcushines) [inline server/entitymanager/entitymanager_test.go:60]: i meant for the format to be applied on all test cases

**Review by marcushines — APPROVED** [review]

**Pull request merged**